src/Framework/PaymentGateways/PaymentGateway.php

```diff
@@ -148,30 +185,118 @@ public function cancelSubscription(Subscription $subscription)
     }
 
     /**
+     * @unreleased
      * @inheritDoc
+     * @throws Exception
      */
     public function canSyncSubscriptionWithPaymentGateway(): bool
     {
-        return $this instanceof SubscriptionTransactionsSynchronizable
-            || $this->subscriptionModule->canSyncSubscriptionWithPaymentGateway();
+        if ($this->subscriptionModule) {
+            return $this->subscriptionModule->canSyncSubscriptionWithPaymentGateway();
+        }
+
+        throw new Exception(esc_html__('Method has not been implemented yet.', 'give'));
     }
 
     /**
+     * @unreleased
      * @inheritDoc
+     * @throws Exception
      */
     public function canUpdateSubscriptionAmount(): bool
     {
-        return $this instanceof SubscriptionAmountEditable
-            || $this->subscriptionModule->canUpdateSubscriptionAmount();
+        if ($this->subscriptionModule) {
+            return $this->subscriptionModule->canUpdateSubscriptionAmount();
+        }
+
+        throw new Exception(esc_html__('Method has not been implemented yet.', 'give'));
     }
 
     /**
+     * @unreleased
      * @inheritDoc
+     * @throws Exception
      */
     public function canUpdateSubscriptionPaymentMethod(): bool
     {
-        return $this instanceof SubscriptionPaymentMethodEditable
-            || $this->subscriptionModule->canUpdateSubscriptionPaymentMethod();
+        if ($this->subscriptionModule) {
+            return $this->subscriptionModule->canUpdateSubscriptionPaymentMethod();
+        }
+
+        throw new Exception(esc_html__('Method has not been implemented yet.', 'give'));
+    }
+
+    /**
+     * @unreleased
+     * @unreleased
+     * @throws Exception
+     */
+    public function hasGatewayDashboardSubscriptionUrl(): bool
+    {
+        if ($this->subscriptionModule) {
+            return $this->subscriptionModule->hasGatewayDashboardSubscriptionUrl();
+        }
+
+        throw new Exception(esc_html__('Method has not been implemented yet.', 'give'));
```

Comment:
Just check if `$this` is an instance of the necessary class. It doesn't need to be manually overriden.

Reply:
@JasonTheAdams As we discussed earlier, `PaymentGateway` should implement all subscription interfaces by default. In this way, I was able to use the gateway function for subscription action instead of accessing them with the subscription module.

I implement a new logic to reduce the chance of overwriting these functions  in child class: https://github.com/impress-org/givewp/compare/7b7148a7802d...f6d72960749d/

Let me know what you think.

Reply:
So, I don't think that works the way you think it would. The `PaymentGateway` class always has those methods, regardless of whether or not the submodule extends them. As such, your utility would always return true (I made a simplified test case to confirm this). Since we're not using any sort of inheritance model, there's no way to use reflection to detect this. But really, the simple answer is that if the gateway doesn't have a subscription module and isn't overriding the `can` method, then it just means it should return false. See 4ec360482.

Reply:
@JasonTheAdams  Reflection class can help to detect whether or not function declared in Child class. If function overwrites in child class that means functionality support by it otherwise not.

Let me know what you think.

https://user-images.githubusercontent.com/1784821/173987825-de74143c-34d8-48a6-83af-42efcc91636c.mov